File: sl/test/Ice/operations/Callback.cs
// **********************************************************************
//
// Copyright (c) 2003-2008 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

using System;
using System.Diagnostics;
using System.Threading;

public class Callback
{
    internal Callback()
    {
        _called = false;
    }

    public virtual bool check()
    {
        lock(this)
        {
            return _called;
        }
    }

    public virtual bool checkWait()
    {
        int cnt = 0;
        do
        {
            lock(this)
            {
                if(_called)
                {
                    return true;
                }
            }
            Thread.Sleep(100);
        }
        while(++cnt < 50);

        return false; // Must be timeout
    }

    public virtual void called()
    {
        lock(this)
        {
            Debug.Assert(!_called);
            _called = true;
        }
    }

    private bool _called;
}

File: java-compat/src/Ice/src/main/java/Ice/Callback.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package Ice;

/**
 * An application can optionally supply an instance of this class in an
 * asynchronous invocation. The application must create a subclass and
 * implement the completed method.
 **/
public abstract class Callback extends IceInternal.CallbackBase
{
    /**
     * Invoked when the invocation completes. The subclass should
     * call the matching <code>end_OP</code> method on the proxy and
     * must be prepared to handle exceptions.
     *
     * @param r The asynchronous result object returned by the <code>begin_OP</code> method.
     **/
    public abstract void completed(AsyncResult r);

    /**
     * Invoked when the Ice run time has passed the outgoing message
     * buffer to the transport. The default implementation does nothing,
     * a subclass can override it if it needs to take action when the
     * message is successfully sent.
     *
     * @param r The asynchronous result object returned by the <code>begin_OP</code> method.
     **/
    public void sent(AsyncResult r)
    {
    }

    @Override
    public final void __completed(AsyncResult r)
    {
        completed(r);
    }

    @Override
    public final void __sent(AsyncResult r)
    {
        sent(r);
    }

    @Override
    public final boolean __hasSentCallback()
    {
        return true;
    }
}
